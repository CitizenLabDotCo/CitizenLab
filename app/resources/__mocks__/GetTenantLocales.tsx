import { GetAppConfigurationLocalesChildProps } from 'resources/GetAppConfigurationLocales';

export const mockGetTenantLocales = ['en', 'fr-BE'] as GetAppConfigurationLocalesChildProps;
