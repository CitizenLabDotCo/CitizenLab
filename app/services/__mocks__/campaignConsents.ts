export const updateConsentByCampaignIDWIthToken = jest.fn().mockImplementation(() => Promise.resolve());
