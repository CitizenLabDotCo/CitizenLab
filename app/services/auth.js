import { API_PATH } from 'containers/App/constants';
import request from 'utils/request';
// import Streams from '../utils/streams';
// import Rx from 'rxjs/Rx';
// import _ from 'lodash';

export function signIn(email, password) {
  const data = {
    auth: {
      email,
      password,
    },
  };

  return request(`${API_PATH}/user_token`, data, {
    method: 'POST',
  });
}

export function observeSignIn(email, password) {
  return Streams.create(`${API_PATH}/ideas/${id}`, null, localProperties);
}