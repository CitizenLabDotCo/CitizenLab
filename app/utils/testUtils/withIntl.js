import { IntlProvider, intlShape } from "react-intl";
import { shallow, mount } from "enzyme";
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider(
  {
    locale: "en",
    messages: {
      message1: "Hello world"
    }
  },
  {}
);
const { intl } = intlProvider.getChildContext();

// `intl` prop is required when using injectIntl HOC
const nodeWithIntlProp = node => React.cloneElement(node, { intl });

// shallow() with React Intl context
export const shallowWithIntl = (node, { context, ...options } = {}) => {
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  });
};

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
