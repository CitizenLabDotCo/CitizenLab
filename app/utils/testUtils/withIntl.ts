import { IntlProvider, intlShape, InjectedIntlProps } from 'react-intl';
import { shallow, mount, ShallowRendererProps, ShallowWrapper } from 'enzyme';
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider({
  locale: 'en',
}, {});

// You customize the intl object here:
const { intl: originalIntl } = intlProvider.getChildContext() as InjectedIntlProps;
const intl = {
  ...originalIntl,
  formatMessage: ({ id, defaultMessage }, values?) =>
    originalIntl.formatMessage({
      id,
      defaultMessage: defaultMessage || id
    }, { tenantName: 'The Test', orgName: 'Test Town', orgType: 'testing', ...values || {} }),
};

const nodeWithIntlProp = (node) => React.cloneElement(node, { intl });

// shallow() with React Intl context
export function shallowWithIntl<C extends React.Component, P = C['props'], S = C['state']>(node: React.ReactElement<P>, additional: ShallowRendererProps = {}) {
 const { context, ...options } = additional;
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  }) as ShallowWrapper<P & InjectedIntlProps, S, C>;
}

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
