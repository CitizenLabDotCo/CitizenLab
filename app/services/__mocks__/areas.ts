import { BehaviorSubject } from 'rxjs';
import { IIdea } from 'services/ideas';

let mockIdea: IIdea | null = null;

export const __setMockIdea = (idea: IIdea) => {
  mockIdea = idea;
};

export const areaByIdStream = jest.fn((_areaId) => {
  const observable = new BehaviorSubject(mockIdea);
  return {
    observable
  };
});
