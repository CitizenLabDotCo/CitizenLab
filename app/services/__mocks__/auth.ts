import { IUser } from 'services/users';
export const mockUser = {
  data: {
    id: '522ae8cc-a5ed-4d31-9aa0-470904934ec6',
    type: 'users',
    attributes: {
      first_name: 'Test',
      last_name: 'Citizenlab',
      slug: 'test-citizenlab',
      locale: 'en',
      avatar: {
        small: null,
        medium: null,
        large: null
      },
      roles: [],
      bio_multiloc: {},
      registration_completed_at: '2018-11-26T15:40:54.355Z',
      invite_status: null,
      created_at: '2018-11-26T15:41:19.782Z',
      updated_at: '2018-11-26T15:41:19.782Z',
      email: 'test@citizenlab.co',
      custom_field_values: {
        birthyear: 1990,
        domicile: 'outside',
        gender: 'female',
        politician: 'retired_politician'
      },
      unread_notifications: 0
    },
    relationships: {
      granted_permissions: {
        data: []
      }
    }
  }
} as IUser;

const adminRole = [{ type: 'admin' }];

function setRoles(roles) {
  const user = mockUser;
  user.data.attributes.roles = roles;
  return user;
}
export const mockAdmin = setRoles(adminRole) as IUser;

export const mockProjectModerator = (projectId: string) => setRoles([{ type: 'project_moderator', project_id: projectId }]) as IUser;
