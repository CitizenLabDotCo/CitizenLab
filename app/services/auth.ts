import { IUser } from 'services/users';
import { IHttpMethod, Locale } from 'typings';
import { API_PATH } from 'containers/App/constants';
import { getJwt, setJwt, removeJwt } from 'utils/auth/jwt';
import request from 'utils/request';
import streams from 'utils/streams';
import clHistory from 'utils/cl-router/history';

export const authApiEndpoint = `${API_PATH}/users/me`;

export interface IUserToken {
  jwt: string;
}

export function authUserStream() {
  return streams.get<IUser | null>({ apiEndpoint: authApiEndpoint });
}

export async function signIn(email: string, password: string) {
  try {
    const bodyData = { auth: { email, password } };
    const httpMethod: IHttpMethod = { method: 'POST' };
    const { jwt } = await request<IUserToken>(`${API_PATH}/user_token`, bodyData, httpMethod, null);
    setJwt(jwt);
    const authenticatedUser = await getAuthUserAsync();
    authUserStream().observer.next(authenticatedUser);
    streams.reset();
    return authenticatedUser;
  } catch (error) {
    signOut();
    throw error;
  }
}

export async function signUp(
  firstName: string,
  lastName: string,
  email: string,
  password: string,
  locale: Locale,
  isInvitation: boolean | null | undefined,
  token: string | undefined | null
) {
  const innerBodyData = {
    email,
    password,
    locale,
    first_name: firstName,
    last_name: lastName
  };

  const httpMethod: IHttpMethod = {
    method: 'POST'
  };

  try {
    const signUpEndpoint = (isInvitation === true ? `${API_PATH}/invites/by_token/${token}/accept` : `${API_PATH}/users`);
    const bodyData = { [token ? 'invite' : 'user']: innerBodyData };
    await request(signUpEndpoint, bodyData, httpMethod, null);
    const authenticatedUser = await signIn(email, password);
    return authenticatedUser;
  } catch (error) {
    throw error;
  }
}

export function signOut() {
  const jwt = getJwt();

  if (jwt) {
    removeJwt();
    authUserStream().observer.next(null);
    streams.reset();

    if (location.pathname.startsWith('/admin')) {
      clHistory.push('/');
    }
  }
}

export async function getAuthUserAsync() {
  try {
    const authenticatedUser = await request<IUser>(authApiEndpoint, null, null, null);
    return authenticatedUser;
  } catch {
    signOut();
    throw new Error('not_authenticated');
  }
}

export async function sendPasswordResetMail(email: string) {
  try {
    const bodyData = {
      user: {
        email
      }
    };
    const httpMethod: IHttpMethod = { method: 'POST' };
    const response = await request(`${API_PATH}/users/reset_password_email`, bodyData, httpMethod, null);
    return response;
  } catch (error) {
    throw error;
  }
}

export async function resetPassword(password: string, token: string) {
  try {
    const bodyData = {
      user: {
        password,
        token
      }
    };
    const httpMethod: IHttpMethod = { method: 'POST' };
    const response = await request(`${API_PATH}/users/reset_password`, bodyData, httpMethod, null);
    return response;
  } catch (error) {
    throw error;
  }
}
