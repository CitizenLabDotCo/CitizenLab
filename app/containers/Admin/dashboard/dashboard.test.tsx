import React from 'react';
import 'jest-styled-components';
jest.mock('utils/cl-intl');
import { DashboardsPage } from './';

import { shallowWithIntl } from 'utils/testUtils/withIntl';
import { shallow } from 'enzyme';
import { mockUser, mockAdmin, mockProjectModerator } from 'services/__mocks__/auth';

describe('<DashboardsPage />', () => {
  it('renders correctly with an admin user', () => {
    const wrapper = shallowWithIntl(<DashboardsPage authUser={mockAdmin.data} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with a project moderator user', () => {
    const wrapper = shallowWithIntl(<DashboardsPage authUser={mockProjectModerator('testProjectId').data} />);
    expect(wrapper).toMatchSnapshot();
  });
});
