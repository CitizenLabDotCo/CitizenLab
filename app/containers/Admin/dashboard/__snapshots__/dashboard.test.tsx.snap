// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DashboardsPage /> renders correctly with a project moderator user 1`] = `
<Component
  onlyModerator={true}
/>
`;

exports[`<DashboardsPage /> renders correctly with an admin user 1`] = `
<withRouter(TabbedResource)
  resource={
    Object {
      "subtitle": "Get immediate and easy-to-grasp analytics into whatâ€™s moving on the platform.",
      "title": "Dashboard",
    }
  }
  tabs={
    Array [
      Object {
        "label": "Summary",
        "url": "/admin/dashboard",
      },
      Object {
        "label": "Users",
        "url": "/admin/dashboard/users",
      },
    ]
  }
>
  <HelmetIntl
    description={
      Object {
        "defaultMessage": "Dashboard for activities on the platform",
        "id": "app.containers.AdminPage.DashboardPage.helmetDescription",
      }
    }
    title={
      Object {
        "defaultMessage": "Admin dashboard page",
        "id": "app.containers.AdminPage.DashboardPage.helmetTitle",
      }
    }
  />
</withRouter(TabbedResource)>
`;
